Ensure large results are handled correctly

alexpeattie opened the pull request:
This PR ensures we first call `read_all_of_child_result_pipe`, before `wait_for_child` which should fix #54.

Although a bit counterintuitive, we should read the data from the pipe before waiting for the child to exit, otherwise when the serialized data exceeds the pipe's buffer capacity, the child process will block indefinitely while trying to write, and the parent process will block indefinitely waiting for the child to exit, resulting in a deadlock.

This actually ends up simplifying our logic: we no longer check `read_all_of_child_result_pipe` for an empty buffer. If the buffer is empty, it means the child process exited unhappily, which will be immediately handled by `error_if_child_unhappy`. On the other hand, if we reach `deserialize_result`, we're guaranteed a non-empty buffer (as the buffer will contain the result of `serialize_result_or_error_value`, which is always non-empty).

I believe the only potential change in behavior is that a pipe read error would now supersede a child process exiting unhappily. I don't think this is a problem in practice.

I also found disabling mocking in the child thread's context seemed to fix #55.

alexpeattie committed:
Ensure large results are handled correctly, fixes #54

brannondorsey commented on src/tests.rs:
```diff
@@ -218,6 +218,16 @@ fn all_function_types() {
     assert_eq!(result, MyResult { value: 5 });
 }
 
+#[rstest]
+#[timeout(TEST_TIMEOUT)]
+fn handle_large_result() {
+    let initial_string = "The quick brown fox jumps over the lazy dog";
+    let result = execute_in_isolated_process(|| {
+        initial_string.repeat(10000)
+    }).unwrap();
+    assert_eq!(result.len(), 10000 * initial_string.len());
+}
+
```
```suggestion
#[rstest]
#[timeout(TEST_TIMEOUT)]
fn handle_large_result() {
    let initial_string = "The quick brown fox jumps over the lazy dog";
    let result = execute_in_isolated_process(|| initial_string.repeat(10_000)).unwrap();
    println!("result.len(): {}", result.len());
    assert_eq!(result.len(), 10_000 * initial_string.len());
}

```

Minor formatting changes to appease `cargo fmt`.

brannondorsey commented on src/lib.rs:
```diff
@@ -321,10 +325,13 @@ where
             close_write_end_of_pipe_in_parent(&sys, write_fd)?;
 
+            // Read the data from the pipe before waiting for the child to exit
+            // to prevent deadlocks.
+            let buffer: Vec<u8> = read_all_of_child_result_pipe(read_fd)?;
+
             let waitpid_bespoke_status = wait_for_child(&sys, child_pid)?;
             error_if_child_unhappy(waitpid_bespoke_status)?;
 
```
This is true, but it isn't clear to me that the older ordering needs to be preserved.

brannondorsey commented on src/lib.rs:
```diff
@@ -554,13 +561,6 @@ fn read_all_of_child_result_pipe(read_fd: c_int) -> Result<Vec<u8>, MemIsolateEr
 
-    if buffer.is_empty() {
-        // TODO: How can we more rigorously know this? Maybe we write to a mem map before and after execution?
-        let err = CallableStatusUnknown(CallableProcessDiedDuringExecution);
```
I believe `CallableProcessDiedDuringExecution` is now an unused import.

brannondorsey reviewed (changes requested):
This is great! Your rationale for reversing the order of the pipe drain and the waitpid makes sense to me. I was starting to explore a solution that would allow us to select/poll on both waitpid and the pipe read and continue processing whichever came in first, but this is much more elegant.

One thing that I'm noticing with this solution in its current state is that it has a different handling of this new test than `main`.

```rs

#[rstest]
#[timeout(TEST_TIMEOUT)]
fn panic_in_child() {
    #[allow(clippy::semicolon_if_nothing_returned)]
    let error = execute_in_isolated_process(|| {
        panic!("Panic in child");
        #[allow(clippy::unused_unit)]
        #[allow(unreachable_code)]
        ()
    })
    .unwrap_err();
    eprintln!("error: {error:?}",);
    assert!(matches!(
        error,
        MemIsolateError::CallableStatusUnknown(
            CallableStatusUnknownError::CallableProcessDiedDuringExecution
        )
    ));
}
```

That makes sense given `CallableProcessDiedDuringExecution` is actually removed, but the new error that returns is unexpected from the perspective of the caller.

```
failures:

---- tests::panic_in_child stdout ----
error: CallableExecuted(DeserializationFailed("io error: unexpected end of file"))

thread 'tests::panic_in_child' panicked at src/tests.rs:242:5:
assertion failed: matches!(error,
    MemIsolateError::CallableStatusUnknown(CallableStatusUnknownError::CallableProcessDiedDuringExecution))


failures:
    tests::panic_in_child

test result: FAILED. 34 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.23s
```

It looks like the `if buffer.is_empty() {}` check that is removed from this PR was actually performing a role, and so perhaps we need to add it back after the `wait_for_child` but before `deserialize_result()`.

I'm actually somewhat surprised to find that `error_if_child_unhappy()` doesn't return a `CallableStatusUnknown(UnexpectedChildExitStatus(unhandled_status))` in the case of a panic inside the closure passed to `execute_in_isolated_process()`. I would have expected that :thinking:.

brannondorsey commented:
> I'm actually somewhat surprised to find that `error_if_child_unhappy()` doesn't return a `CallableStatusUnknown(UnexpectedChildExitStatus(unhandled_status))` in the case of a panic inside the closure passed to `execute_in_isolated_process()`. I would have expected that :thinking:.

If I remove the existing `if buffer.is_empty()` check in `main` and run with debug tracing on, I see the same error.

```
RUST_LOG=debug cargo test tests::panic_in_child --features tracing
```

```
2025-04-22T02:38:07.098090Z DEBUG execute_in_isolated_process:parent:get_system_functions: mem_isolate: using MockableSystemFunctions
2025-04-22T02:38:07.098220Z DEBUG execute_in_isolated_process:parent:create_pipe{sys=MockableSystemFunctions}: mem_isolate: pipe created: PipeFds { read_fd: 3, write_fd: 4 }
2025-04-22T02:38:07.098876Z DEBUG execute_in_isolated_process:parent:close_write_end_of_pipe_in_parent{sys=MockableSystemFunctions write_fd=4}: mem_isolate: write end of pipe closed successfully
2025-04-22T02:38:07.098989Z DEBUG execute_in_isolated_process:parent:wait_for_child{sys=MockableSystemFunctions child_pid=74541}: mem_isolate: waiting for child process
2025-04-22T02:38:07.099371Z DEBUG execute_in_isolated_process:child:close_read_end_of_pipe_in_child_or_exit{sys=MockableSystemFunctions writer=File { fd: 4, path: "pipe:[286250]", read: false, write: true } read_fd=3}: mem_isolate: read end of pipe closed successfully
2025-04-22T02:38:07.099470Z DEBUG execute_in_isolated_process:child:execute_callable: mem_isolate: starting execution of user-supplied callable
2025-04-22T02:38:07.100732Z DEBUG execute_in_isolated_process:parent:wait_for_child{sys=MockableSystemFunctions child_pid=74541}: mem_isolate: wait completed, received status: 0
2025-04-22T02:38:07.100859Z DEBUG execute_in_isolated_process:parent:error_if_child_unhappy{waitpid_bespoke_status=0}: mem_isolate: child process exited happily on its own
2025-04-22T02:38:07.100951Z DEBUG execute_in_isolated_process:parent:read_all_of_child_result_pipe{read_fd=3}: mem_isolate: successfully read 0 bytes from pipe
2025-04-22T02:38:07.101085Z ERROR execute_in_isolated_process:parent:deserialize_result{buffer=[]}: mem_isolate: failed to deserialize result, propagating CallableExecuted(DeserializationFailed("io error: unexpected end of file"))
test tests::panic_in_child ... FAILED
```

These lines in particular are interesting...

```
execute_in_isolated_process:child:execute_callable: mem_isolate: starting execution of user-supplied callable
execute_in_isolated_process:parent:wait_for_child{sys=MockableSystemFunctions child_pid=74541}: mem_isolate: wait completed, received status: 0
execute_in_isolated_process:parent:error_if_child_unhappy{waitpid_bespoke_status=0}: mem_isolate: child process exited happily on its own
```

Somehow a `panic!()` in the child process is still causing `waitpid` to receive a status of `0` and incorrectly report `child process exited happily on its own`. The empty buffer check is what seems to be protecting that from being incorrectly communicating that up to the caller of `execute_in_isolated_process()`.

brannondorsey commented:
I've just added the test that's mentioned above to `main` in https://github.com/brannondorsey/mem-isolate/pull/57.

I think the lint changes + re-including the `if buffer.is_empty() {}` block as it previously existed to now live right before `deserialize_result(&buffer)` should be enough to get this merged.

> Somehow a panic!() in the child process is still causing waitpid to receive a status of 0 and incorrectly report child process exited happily on its own. The empty buffer check is what seems to be protecting that from being incorrectly communicated up to the caller of execute_in_isolated_process().

I still want to do a bit more digging on this, but I don't mean to block your PR on it. I'm also happy to open a PR based off your branch which includes these changes and perform a regular merge to `main` (not a squash merge) so that this PR gets merged in the process and you `git` the credit ;)

brannondorsey committed:
chore: Add latest changes from main

brannondorsey committed:
chore: Add debug! statement

brannondorsey committed:
fix: Preserve the CallableProcessDiedDuringExecution error from before pipe read() and waitpid() order swap

brannondorsey committed:
Merge branch 'main' into large-result-fix

brannondorsey commented:
> I'm also happy to open a PR based off your branch which includes these changes and perform a regular merge to main (not a squash merge) so that this PR gets merged in the process and you git the credit ;)

@alexpeattie I didn't realize that I could actually commit straight to this branch. I hope that's alright. Let me know what you think of these changes, and if they LGTY I think we can go ahead and merge.

brannondorsey committed:
chore: Rename new buffer empty check function to align better with existing code

alexpeattie commented:
Thanks for your help on this Brannon!

Unfortunately, experimenting more, I think the rabbit-hole goes deeper 😅...

On both `main` and this PR's branch, it seems we actually have a difference between the test and non-test environment.

For instance, create `src/main.rs` with this contents:

```rust
use mem_isolate::{execute_in_isolated_process, MemIsolateError, errors::CallableStatusUnknownError};

fn main() {
    #[allow(clippy::semicolon_if_nothing_returned)]
    let _error = execute_in_isolated_process(|| {
        panic!("Panic in child");
        #[allow(clippy::unused_unit)]
        #[allow(unreachable_code)]
        ()
    });
    let error = _error.unwrap_err();

    assert!(matches!(error,
        // 101 because:
        // https://github.com/rust-lang/rust/blob/1.86.0/library/core/src/macros/panic.md?plain=1#L65-L66
        MemIsolateError::CallableStatusUnknown(
            CallableStatusUnknownError::UnexpectedChildExitStatus(101)
        )
    ), "Expected UnexpectedChildExitStatus(101) but got {:?}", error);
    println!("Success!");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_main() {
        main();
    }
}
```

For me, running `cargo run` I get, on both macOS and Linux:

- `"Success!"` printed: the assertion passed, and the panic was caught. On both OSes, `error` is consistently `UnexpectedChildExitStatus(101)`
- However, I get output as if there was a fatal panic, including a backtrace with `RUST_BACKTRACE=1 cargo run` 

But, on both OSes, running `test_main` (e.g. via `cargo test -- tests::test_main --exact --show-output`) **fails** but differently for macOS and Linux. In Linux I get:

```
thread 'tests::test_main' panicked at src/main.rs:13:5:
Expected UnexpectedChildExitStatus(101) but got CallableStatusUnknown(CallableProcessDiedDuringExecution)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

On macOS I get:

```
thread 'tests::test_main' panicked at src/main.rs:13:5:
Expected UnexpectedChildExitStatus(101) but got CallableStatusUnknown(ChildProcessKilledBySignal(5))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Are you able to reproduce?

I think the reason the test environment behaves a differently is that the `test` crate wraps each test function in a `panic::catch_unwind` (see [here](https://github.com/rust-lang/rust/blob/e3d68139200da0e5010321b286e3e15e4a9530e0/src/tools/compiletest/src/executor.rs#L124)) and there's some strange interactions going on because of this. I'd need to dig more to understand why the cross-platform difference is happening.

Basically, I think it's not safe to `panic` in the child and let that propagate - as it leads to these quirks (like seeing a backtrace outside of test env, and the quirky behavior in tests). Instead, I think `execute_callable` should be wrapped in our own `panic::catch_unwind` so we force the behavior we want across all OSes and test/non-test env. That should also get rid of the need for `error_if_buffer_is_empty`.

I can take a stab at this on the weekend if you don't get there first 🙂

alexpeattie commented:
(I think AOK to merge this in the meantime if you'd like, as it still fixes the original issue 👍)

brannondorsey committed:
fix: Prevent creating zombies by erroring before calling waitpid

If a child process exits and the parent is still alive, the child
becomes a zombie so the parent still has a chance to wait() for it.

Rearranging the pipe result error to be returned after the wait does two
things:

1. It prevents these zombies from being created
2. It preserves the original error order/behavior before the deadlock
fix

brannondorsey commented:
> Are you able to reproduce?

Yep, I just reproduced exactly the scenario you describe on Linux. I'm not at a MacOS machine at the moment, but I encountered the same exact behavior you describe in the test environment [via this GitHub action I added last night](https://github.com/brannondorsey/mem-isolate/pull/60/files#r2055195074), so I'm pretty confident our understanding of the behavior on MacOS is correct.

That case is particularly intriguing because signal 5 is a SIGTRAP :thinking:, which is a bit befuddling to me.

> I think the reason the test environment behaves a differently is that the test crate wraps each test function in a panic::catch_unwind (see [here](https://github.com/rust-lang/rust/blob/e3d68139200da0e5010321b286e3e15e4a9530e0/src/tools/compiletest/src/executor.rs#L124)) and there's some strange interactions going on because of this. I'd need to dig more to understand why the cross-platform difference is happening.
>
> Basically, I think it's not safe to panic in the child and let that propagate - as it leads to these quirks (like seeing a backtrace outside of test env, and the quirky behavior in tests). Instead, I think execute_callable should be wrapped in our own panic::catch_unwind so we force the behavior we want across all OSes and test/non-test env. That should also get rid of the need for error_if_buffer_is_empty.

This is indeed an interesting find! It aligns with what I've been seeing with the different behavior across OSes, but I hadn't yet noticed the behavior was different in test and non-test environments as well.

> I can take a stab at this on the weekend if you don't get there first 🙂

Lovely! I too have only limited availability during the work week to hack on this stuff. I'm curious to hear what you find, and I'll also share anything interesting I may learn about this scenario if I do get around to exploring more before you.

> (I think AOK to merge this in the meantime if you'd like, as it still fixes the original issue 👍)

Sounds good. I just pushed one final behavior fix https://github.com/brannondorsey/mem-isolate/pull/56/commits/ef5650379b30fc8b09aedd1359609a70b185f256 to this PR tracking the original issue. I'll merge this now and we can track the other scenario you've discovered as a separate issue.

Thanks again for the help and the interesting finds you're producing :smile_cat:

brannondorsey commented:
FYI I've [cut a new release](https://github.com/brannondorsey/mem-isolate/releases/tag/v0.1.6) that includes these changes, and I've also opened https://github.com/brannondorsey/mem-isolate/pull/61 in case its helpful as we keep digging on the MacOS/test discrepancies. 

I haven't yet opened a ticket for that one, so feel free to do that, or else I can probably this weekend.
